### PR TITLE
MLX: escape sampled text, join the rank consensus, notice a missing eval cadence

### PR DESCRIPTION
**Pull request by danielhanchen**
Three follow-ups to #1011, which merged before these could go onto its branch. Each is pinned by a test that was mutation-checked: revert the fix and exactly that test fails.

## Sampled text could drive the terminal

`generate_during_eval` prints held-out prompts and model completions. Prompts come from the dataset and completions from the model, so both are untrusted text on its way to a terminal, and `_one_line` collapsed whitespace with `str.split()`, which strips whitespace only:

```
_one_line("safe\x1b]52;c;U0VDUkVU\x07forged", 96)
-> "safe\x1b]52;c;U0VDUkVU\x07forged"
```

ESC, BEL and the rest of C0/C1 survive, so an OSC sequence reaches the clipboard and a CSI sequence erases the lines above it, leaving the surrounding training log reading as something it is not. U+202E reverses the apparent order of everything after it. Split labels bypassed `_one_line` entirely, so a dict eval dataset printed its key raw.

Only `Cc`/`Cf` are escaped, so CJK, emoji, combining marks and backslashes are unchanged, which `repr()` would not have given us.

## A failing decode left its peers at the collective

`_sample_generations` is wrapped, which keeps a raise from the engine off the run. The remaining half is that a rank which unwinds never reaches the `_distributed_should_stop()` call between the two decodes, so a healthy peer waits there for a partner that has already left. Referenced DPO is the configuration that reaches it, since it is the one that decodes twice.

Both decodes now join a rank-wide consensus and skip together, the shape `_evaluate_batch_totals` already uses for this hazard. Interrupts still propagate unwrapped.

## Best-model selection with no evaluation cadence was silent

With an eval dataset, `load_best_model_at_end=True` and the default `eval_steps=0`, nothing raises an evaluation: `_run_best_tracking` never runs, `_best_step` stays `None`, and the restore is guarded on exactly that. Best-model loading was asked for and silently not done. A callback may legitimately own the cadence, so this notices rather than refuses.

The same helper fixes the neighbouring `generate_during_eval` notice, which reported `eval_strategy="epoch"` as no cadence because it tested `eval_steps` alone.

## Verification

- 301 tests across `tests/test_mlx_preference.py` and `tests/test_mlx_trainer_internals.py`.
- 39 MLX test files swept, no failures.
- A compatibility matrix run twice, 14/14 both times: in a venv with no mlx installed, where `is_mlx_available()` must stay False and the MLX modules must not load, and against real mlx 0.32.2 through the `mlx-cpu` manylinux wheel. It covers the pre-PR config payload round-trip, the `MLXTrainingConfig(**kwargs)` call shape Studio uses on Apple Silicon, the best-metric direction, and CJK/emoji/backslash passthrough. Without this change the matrix fails its terminal check.

One note, not a defect. The `self.model.eval()` around `_sample_generations` looks redundant: `generate_batch` already enters `generation_mode(model)`, which snapshots the training flags, calls `model.eval()` and restores them in a `finally` (`unsloth_zoo/mlx/generate.py:555-600`, entered at `generate.py:1702`). Harmless as belt and braces, but worth knowing if it was thought to be load bearing.

### CODE REVIEW

**Comment by danielhanchen** [comment]
Added a fourth fix, and the evidence for the third one is now from a real MLX run rather than the torch shim.

## Fourth fix: an eval split is held to the length it declares

Validation calls `len()` on each eval split and treats it as the guarantee the pass ends, but `create_preference_batch_plan` does a bare `for raw in dataset:` and is never bounded by it. The check declares a length without holding the split to one.

Confirmed on real mlx 0.32.2 with `mlx-community/SmolLM-135M-Instruct-4bit`: a split reporting `len() == 2` and yielding 3 rows had `len()` called exactly once, by the validator, and all 3 rows were formatted, batched, scored and sampled. A sized but endless iterable hangs at the first evaluation for the same reason, which is the outcome the check exists to prevent.

The bound is in `_format_preference_split`, which wraps the eval path only, so training is untouched.

## The generation-failure fix, measured on real MLX

The earlier evidence for the sampling guard was from the torch shim. Re-run against real mlx on the pre-fix tree, ORPO with evaluations at steps 2, 4 and 6 and the engine raising at the second evaluation:

- healthy: reaches step 6, `on_log(eval)` and `on_evaluate` fire at 2, 4 and 6, three eval entries in `log_history`.
- failing: the run dies at step 4 of 6. `_last_eval_metrics` holds a complete, valid 12-key metric set with `eval_loss` 11.496148, but `log_history` contains only the step-2 entry and `state.best_metric` is stuck at step 2's 11.502584 despite 11.496148 being better, so best tracking never ran for an evaluation that had already succeeded.
- `last_generation_samples` afterwards still held the stale step-2 samples, so a consumer reading it after the crash sees samples from the wrong step. The fix sets it to `[]`.

## Also verified on real MLX, no change needed

- The eval dataset is read exactly once per run, across three evaluations: `__iter__` 1, one `formatting_func` call per row, no re-render. The prompt in `last_generation_samples` is byte-identical to the string the scorer rendered, checked with a deliberately stateful `apply_chat_template` that stamps a serial on every render, so a second render would have shifted it.
- The repo's real-MLX Linux gate is undisturbed: `test_mlx_rng_rewind`, `test_mlx_quantized_optimizer_state`, `test_mlx_shape_guard`, `test_mlx_vlm_label_masks` and `test_mlx_batching_and_decay` give 411 passed, 0 failed, 0 skipped on this branch against 403 on the merge base, the difference being the 8 tests the branch adds.

302 tests now pass across the two suites, and each fix here is pinned by a test that was mutation-checked.

What this host cannot answer: everything above is single-rank CPU, so the distributed consensus path added here and any Metal-specific generation behaviour are unverified. Those need an Apple Silicon runner and a multi-rank run.

File: unsloth_zoo/mlx/trainer.py [hunk 1]
```diff
@@ -7788,6 +7850,12 @@ def _prepare_data(self, is_vlm, defer_vlm_checker=False):
             # A cadence is optional -- a callback can raise should_evaluate --
             # but selecting a best model reads a metric only an evaluation makes.
             _sampling_eval = bool(getattr(args, "generate_during_eval", False))
+            # An epoch strategy is a cadence too, and is resolved separately
+            # from eval_steps, so testing eval_steps alone misreports it.
+            _eval_cadence = (
+                _resolve_interval_steps(args.eval_steps, 1) > 0
+                or self._epoch_cadence_enabled("eval_strategy")
+            )
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/trainer.py:7883]: Correct, and fixed in 68aa4243. Confirmed the premise: _static_cadence_enabled returns False for "no", and the eval trigger requires _static_eval_cadence_enabled(), so eval_steps > 0 under "no" evaluates never; a probe run reported no eval_loss in log_history and _best_step None while the notice stayed silent. The inconsistency was mine: the same expression already honoured a hand-set "epoch" through _epoch_cadence_enabled on that field, so ignoring a hand-set "no" read only half of it. The step branch now gates on _static_eval_cadence_enabled() too, which is a no-op for the derived "steps" value and for a missing field.

File: unsloth_zoo/mlx/trainer.py [hunk 2]
```diff
@@ -6387,19 +6410,53 @@ def _sample_generations(current_step):
...
                 modules = tuple(getattr(_sampling_reference, "modules", ()) or ())
                 if modules and not self._distributed_should_stop():
                     scales = [module.scale for module in modules]
                     try:
                         for module in modules:
                             module.scale = 0.0
-                        reference = generate_batch(
-                            model, self.tokenizer, requests, defaults=defaults,
-                        )
+                        reference = _decode("reference")
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/trainer.py:6476]: Correct, and fixed in 68aa4243. Reproduced at head: the reference path printed "reference generation failed (...); skipping samples for this evaluation" and then published two rows with reference None, which is byte-identical to what test_unreferenced_objectives_sample_the_policy_only asserts a legitimate reference-free run publishes, so the log line was false and the failure invisible. The clear-and-return now sits after the finally that restores the adapter scales and inside the modules branch, so reference-free and distributed-stop paths are untouched. Pinned by test_a_failed_reference_decode_publishes_no_samples, which fails without the fix.

File: unsloth_zoo/mlx/trainer.py [hunk 3]
```diff
@@ -938,8 +939,30 @@ def _build_generation_defaults(args):
...
+
+    Prompts come from the dataset and completions from the model, and
+    str.split() strips only whitespace: ESC/BEL survive, so an OSC sequence
+    reaches the clipboard and a CSI one rewrites the log above. Only Cc/Cf are
+    escaped, so CJK, emoji and backslashes stay as they are.
+    """
+    return "".join(
+        (
+            character
+            if unicodedata.category(character) not in ("Cc", "Cf")
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/trainer.py:952]: Not taking this one, though the Unicode facts are right: U+200D and the U+E0020-U+E007F tag block are both Cf and are escaped, and U+FE0F is Mn so it already passes through. Two reasons. The effect is display only, in a preview that _one_line already truncates at 96 characters; _escape_terminal_controls is applied at the print sites and never to last_generation_samples, so no consumed value changes. More importantly the proposed narrowing to the bidi set would reopen a real vector: the tag block is the carrier for invisible-text smuggling, and a completion carrying 16 hidden tag characters renders as clean text once the block is exempted, in exactly the surface an operator reads model output from. Escaping all of Cc/Cf is the conservative choice. The docstring did overstate this, so 68aa4243 corrects it to say emoji survive unless they join with U+200D, and why the tag block stays escaped.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
Settling the note in the description about `self.model.eval()` around `_sample_generations`. It is measured now rather than suspected, on real mlx 0.32.2 with `mlx-community/SmolLM-135M-Instruct-4bit` and a second real `mlx_lm.models.llama`, reading the embedding's own `training` flag inside every forward.

Suppressing only that call, leaving `generation_mode()` untouched, changes nothing: 0 of 5 generation forwards saw `training=True` on SmolLM, 0 of 20 on the second model, identical to baseline. The instrument is not blind, which is the part that matters: a positive control that also suppresses `generation_mode()`'s own eval shows 5 of 5 and 20 of 20 forwards with `training=True` and NEFTune noise applied.

So `generation_mode(model)` is already what keeps NEFTune and dropout out of the decode, and the added call is belt and braces. It is paired with a `finally` and no behavioural difference showed up, so there is nothing to change here; recording it so the next reader does not assume it is load bearing.

Two related results from the same runs, both on the code that merged in #1011 rather than on this PR:

The eval-mode gate inside `_evaluate` is load bearing. Removing `self.model.eval()` there puts NEFTune noise into the eval forward and moves the number: `eval_loss` 3.19348 to 3.22053 on the 4-bit model.

Removing the paired `finally: self.model.train()` is not observable on the happy path, because `_run_eval` calls `model.train()` itself on the line after `_evaluate` returns. The flag does differ in the window the `finally` owns, and the `finally` is load bearing on the exception path: with it intact an injected raise still leaves training mode restored, with it removed the model is left in eval mode. Worth knowing for anyone writing a test against that pairing.

**Comment by danielhanchen** [comment]
CI on this branch is blocked by a failure that predates it. `MLX (Linux CPU: numerics + suites)` fails at `tests/test_mlx_neftune_quant_map.py (mlx_only half, HARD GATE)`, and the same job fails the same step on `main` itself at 91383f0f, so it is not caused by anything here.

Root cause and fix are in #1125: `_ModelingLoader` forwards only `create_module` and `exec_module`, so `get_source` and `get_code` vanish once the fused-forward finder is installed, and `_transformers_scales_inside_embedding` returns `None` instead of reading transformers through the loader as it is meant to.

The other red checks on this branch were runs cancelled mid-flight, not failures; `gh` reports a cancelled run as `fail`. Those have been re-run.

This branch's own tests are green: 80 in `tests/test_mlx_preference.py` and 226 in `tests/test_mlx_trainer_internals.py`, with every fix here mutation-checked.
